File: backend/services/intent_service.py
import re

def detect_intent(message: str):
    query = message.lower()
    
    rules = [
        {"intent": "greeting", "keywords": [r"hi", r"hello", r"hey", r"greetings", r"hola", r"help", r"start"]},
        {"intent": "registration", "keywords": [r"regist", r"apply", r"sign up", r"form 6", r"enroll", r"eligib"]},
        {"intent": "voting", "keywords": [r"vote", r"voting", r"ballot", r"booth", r"cast", r"evm", r"vvpat"]},
        {"intent": "documents", "keywords": [r"id ", r"document", r"paper", r"proof", r"aadhaar", r"epic", r"licen"]},
        {"intent": "timeline", "keywords": [r"date", r"when", r"deadline", r"timeline", r"schedule", r"phase"]},
        {"intent": "results", "keywords": [r"result", r"count", r"winner", r"who won", r"leads"]},
        {"intent": "location", "keywords": [r"poll", r"station", r"site", r"map", r"where", r"locat"]}
    ]

    for rule in rules:
        matches = [k for k in rule["keywords"] if re.search(k, query)]
        if matches:
            confidence = min(0.5 + (len(matches) * 0.1), 0.95)
            return {"intent": rule["intent"], "confidence": confidence}

    return {"intent": "general", "confidence": 0.1}

File: backend/src/services/intentService.js
/**
 * Intent Detection Service
 * Categorizes user queries based on keywords to route to the correct local knowledge logic.
 *
 * @param {string} message - The raw text input from the user.
 * @returns {{ intent: string, confidence: number }} An object containing the matched intent and a confidence score.
 */
export const detectIntent = (message) => {
  const query = message.toLowerCase();
  
  const rules = [
    { intent: 'registration', keywords: [/regist/i, /apply/i, /sign up/i, /form 6/i, /enroll/i] },
    { intent: 'voting', keywords: [/vote/i, /voting/i, /ballot/i, /booth/i, /cast/i, /evm/i, /vvpat/i] },
    { intent: 'documents', keywords: [/id /i, /document/i, /paper/i, /proof/i, /aadhaar/i, /epic/i, /licen/i] },
    { intent: 'timeline', keywords: [/date/i, /when/i, /deadline/i, /timeline/i, /schedule/i, /phase/i] },
    { intent: 'results', keywords: [/result/i, /count/i, /winner/i, /who won/i, /leads/i] }
  ];

  for (const rule of rules) {
    const matchCount = rule.keywords.filter(k => k.test(query)).length;
    if (matchCount > 0) {
      // Calculate a basic confidence score
      const confidence = Math.min(0.5 + (matchCount * 0.1), 0.95);
      return { intent: rule.intent, confidence };
    }
  }

  return { intent: 'general', confidence: 0.1 };
};
